packages/core/src/render3/i18n.ts

```diff
@@ -505,43 +513,81 @@ function appendI18nNode(tNode: TNode, parentTNode: TNode, previousTNode: TNode |
  */
 export function i18nPostprocess(
     message: string, replacements: {[key: string]: (string | string[])} = {}): string {
-  //
-  // Step 1: resolve all multi-value cases (like [�*1:1��#2:1�|�#4:1�|�5�])
-  //
-  const matches: {[key: string]: string[]} = {};
-  let result = message.replace(PP_PLACEHOLDERS, (_match, content: string): string => {
-    if (!matches[content]) {
-      matches[content] = content.split('|');
-    }
-    if (!matches[content].length) {
-      throw new Error(`i18n postprocess: unmatched placeholder - ${content}`);
+  /**
+   * Step 1: resolve all multi-value placeholders like [�#5�|�*1:1��#2:1�|�#4:1�]
+   *
+   * Note: due to the way we process nested templates (BFS), multi-value placeholders are typically
+   * grouped by templates, for example: [�#5�|�#6�|�#1:1�|�#3:2�] where �#5� and �#6� belong to root
+   * template, �#1:1� belong to nested template with index 1 and �#1:2� - nested template with index
+   * 3. However in real templates the order might be different: i.e. �#1:1� and/or �#3:2� may go in
+   * front of �#6�. The post processing step restores the right order by keeping track of the
+   * template id stack and looks for placeholders that belong to the currently active template.
+   */
+  let result: string = message;
+  if (PP_MULTI_VALUE_PLACEHOLDERS_REGEXP.test(message)) {
+    const matches: {[key: string]: PostprocessPlaceholder[]} = {};
+    const templateIdsStack: number[] = [ROOT_TEMPLATE_ID];
+    result = result.replace(PP_PLACEHOLDERS_REGEXP, (m: any, phs: string, tmpl: string): string => {
+      const content = phs || tmpl;
+      if (!matches[content]) {
+        const placeholders: PostprocessPlaceholder[] = [];
+        content.split('|').forEach((placeholder: string) => {
+          const match = placeholder.match(PP_TEMPLATE_ID_REGEXP);
+          const templateId = match ? parseInt(match[1], 10) : ROOT_TEMPLATE_ID;
+          const isCloseTemplateTag = PP_CLOSE_TEMPLATE_REGEXP.test(placeholder);
+          placeholders.push([templateId, isCloseTemplateTag, placeholder]);
+        });
+        matches[content] = placeholders;
```

Comment:
Nit: I think this block can be broken out into its own function for readability
```suggestion
        matches[content] = getPlaceholders(...);
```

Reply:
Thanks for the review @kara. I think we'll need to refactor this code later (as a part of i18n service work) and move more logic to the compiler, so I'll keep this as is for now.